### PR TITLE
add seaport deployer 

### LINKED ISSUE #16: Deploy Seaport to the era-test-node

### PR DESCRIPTION
**Pull request by daphneherlambda**
Closes #16 
To test this run the deploy command to run the seaport-deployer script `yarn hardhat deploy-zksync --script seaport-deployer.ts
`

### CODE REVIEW

File: deploy/seaport-deployer.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  console.log('provider');
+  const provider = await getProvider();
+  console.log('transaction');
+  const transaction = await provider.getTransaction(transferHelper.deployTransaction.hash);
+  console.log('code');
+  const creationCode = transaction.data;
+  console.log('safecreate');
+  
+  const transferHelper_address = await immutableCreate2.safeCreate2(bytesFormat , creationCode, {gasLimit: 1000000}); 
+  console.log('Result:', transferHelper_address);
```
Comment (fkrause98) [inline deploy/seaport-deployer.ts:41]: Remove these console.logs please :)

File: contracts/interfaces/ImmutableCreate2FactoryInterface.sol [hunk 2]
```diff
@@ -1,96 +1,204 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.8.13;
+pragma solidity 0.8.17;
```
Comment (fkrause98) [inline contracts/interfaces/ImmutableCreate2FactoryInterface.sol:2]: Maybe we're missing a caret here for the version number?

File: Makefile [hunk 3]
```diff
@@ -43,3 +43,13 @@ compile-seaport: compile-and-deploy-execution-helper
 .PHONY: run-era-test-node
 run-era-test-node: era-test-node
 	cd era-test-node && cargo +nightly run -- --show-calls=all --resolve-hashes run
+
+# ------------------------------------------------------------------------------
+# Clean:
+# ------------------------------------------------------------------------------
+
+.PHONY: clean-execution-helper
+clean-execution-helper: cd ExecutionHelper && yarn hardhat clean && yarn cache clean
```
Comment (fkrause98) [inline Makefile:52]: Add a newline before this command:
```suggestion
clean-execution-helper: 
	cd ExecutionHelper && yarn hardhat clean && yarn cache clean
```

File: deploy/seaport-deployer.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  
+  console.log('provider');
+  const provider = await getProvider();
+  console.log('transaction');
+  const transaction = await provider.getTransaction(transferHelper.deployTransaction.hash);
+  console.log('code');
+  const creationCode = transaction.data;
+  console.log('safecreate');
+  
+  const transferHelper_address = await immutableCreate2.safeCreate2(bytesFormat , creationCode, {gasLimit: 1000000});
```
Comment (fkrause98) [inline deploy/seaport-deployer.ts:40]: Can we open an issue on the gasLimit to be used or do it properly? Maybe we should discuss this with @ilitteri.

**Review by fkrause98 — COMMENTED** [review]
Just double checking, got this output with these changes:
```
> yarn hardhat deploy-zksync --script deploy/seaport-deployer.ts --network inMemoryNode
yarn run v1.22.19
$ /Users/fran/Programming/zksync_seaport/node_modules/.bin/hardhat deploy-zksync --script deploy/seaport-deployer.ts --network inMemoryNode

Starting deployment process of "ImmutableCreate2Factory"...
Estimated deployment cost: 0.00042753025 ETH

"ImmutableCreate2Factory" was successfully deployed:
 - Contract address: 0xCEb687451f3ed6f06B9F4a9E7eca803849534314
 - Contract source: contracts/interfaces/ImmutableCreate2FactoryInterface.sol:ImmutableCreate2Factory - Encoded constructor arguments: 0x


Starting deployment process of "ConduitController"...
Estimated deployment cost: 0.00114059775 ETH

"ConduitController" was successfully deployed:
 - Contract address: 0xD59A4813e5A847DE97785F40f37BCeb58f86eE84
 - Contract source: seaport-core/src/conduit/ConduitController.sol:ConduitController
 - Encoded constructor arguments: 0x


Starting deployment process of "Conduit"...
Estimated deployment cost: 0.00054577275 ETH

"Conduit" was successfully deployed:
 - Contract address: 0x7aCF93D0840fC41D77b3e7498b419FF0eE3a583A
 - Contract source: seaport-core/src/conduit/Conduit.sol:Conduit
 - Encoded constructor arguments: 0x


Starting deployment process of "Seaport"...
Estimated deployment cost: 0.0017423935 ETH

"Seaport" was successfully deployed:
 - Contract address: 0x2763cDcAC15fC3ed254b089cA2a964b9688Eee78
 - Contract source: contracts/Seaport.sol:Seaport
 - Encoded constructor arguments: 0x000000000000000000000000d59a4813e5a847de97785f40f37bceb58f86ee84


Starting deployment process of "TransferHelper"...
Estimated deployment cost: 0.00078561325 ETH

"TransferHelper" was successfully deployed:
 - Contract address: 0x1bD92475184E805e3d0B45E768015A4111fB3E2d
 - Contract source: contracts/helpers/TransferHelper.sol:TransferHelper
 - Encoded constructor arguments: 0x000000000000000000000000d59a4813e5a847de97785f40f37bceb58f86ee84

✨  Done in 2.59s.
```


Let's address the comments I've left before merging 👍, and great work @daphneherlambda 🚀 !

File: deploy/seaport-deployer.ts (hunk 1 above)
Comment (jpcenteno) [reply deploy/seaport-deployer.ts]: Done

File: contracts/interfaces/ImmutableCreate2FactoryInterface.sol (hunk 2 above)
Comment (jpcenteno) [reply contracts/interfaces/ImmutableCreate2FactoryInterface.sol]: What's the probllem?

File: Makefile (hunk 3 above)
Comment (jpcenteno) [reply Makefile]: Resolved

**Comment by jpcenteno** [comment]
Closes #17 
Closes #22 

Does this work?

File: deploy/seaport-deployer.ts (hunk 4 above)
Comment (jpcenteno) [reply deploy/seaport-deployer.ts:35]: Created #27 to track this one.

File: Makefile [hunk 5]
```diff
@@ -43,3 +46,39 @@ compile-seaport: compile-and-deploy-execution-helper
 .PHONY: run-era-test-node
 run-era-test-node: era-test-node
 	cd era-test-node && cargo +nightly run -- --show-calls=all --resolve-hashes run
+
+env:
+	source .env
```
[the remark this answers is not in the document]
Comment (jpcenteno) [reply Makefile:51]: I wasn't even expecting this error

```sh
❯ make env
source .env
make: source: No such file or directory
make: *** [env] Error 1
```

[the remark this answers is not in the document]
Comment (jpcenteno) [reply Makefile:51]: The file exists

```sh
❯ test -f .env; echo $?
0
```

Comment (jpcenteno) [inline Makefile:51]: My humble suggestion

```suggestion
```

File: Makefile [hunk 6]
```diff
@@ -1,3 +1,5 @@
+.PHONY: clean env deploy-execution-helper deploy-seaport deploy
```
[the remark this answers is not in the document]
Comment (jpcenteno) [reply Makefile:1]: `setup:` is also missing the `.PHONY` attribute

**Review by jpcenteno — CHANGES_REQUESTED** [review]

File: Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpcenteno) [reply Makefile]: Fixed d87c7f9

File: Makefile [hunk 7]
```diff
@@ -43,3 +46,39 @@ compile-seaport: compile-and-deploy-execution-helper
...
+# Clean:
+# ------------------------------------------------------------------------------
+
+.PHONY: clean-execution-helper
+clean-execution-helper: 
+	cd ExecutionHelper && \
+	yarn hardhat clean && \
+	yarn cache clean
+
+clean-seaport:
```
[the remark this answers is not in the document]
Comment (jpcenteno) [reply Makefile:73]: Fixed d87c7f9

File: Makefile [hunk 8]
```diff
@@ -43,3 +46,39 @@ compile-seaport: compile-and-deploy-execution-helper
...
+clean: clean-execution-helper clean-seaport
+	yarn hardhat clean
+
+# ------------------------------------------------------------------------------
+# Deploy:
+# ------------------------------------------------------------------------------
+
+# The name of this target is good for now. If in the future we'd add more libs
+# this should be a general target for deploying all of them befor deploying the
+# concret project.
```
[the remark this answers is not in the document]
Comment (jpcenteno) [reply Makefile:87]: Fixed d87c7f9

File: contracts/interfaces/ImmutableCreate2FactoryInterface.sol (hunk 2 above)
Comment (daphneherlambda) [reply contracts/interfaces/ImmutableCreate2FactoryInterface.sol]: its using the same version as the seaport.sol contract, to keep it uniform

File: .env
[the remark this answers is not in the document]
Comment (daphneherlambda) [reply .env:1]: as we are hardcoding the privatekey to deploy to use the same rich wallet, its not convenient to have it as .env.example, but maybe if later we make this to be configurable by the user we can change it

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (jpcenteno) [reply Makefile]: Resolved @ 5f7ab96

File: .env
[the remark this answers is not in the document]
Comment (jpcenteno) [reply .env:1]: I got ahead of myself and changed that @ 5f7ab96eb6053bfec39ff8b253f7406f138fe030. reverted @ 3edd2c7

**Review by fkrause98 — APPROVED** [review]

**Review by jpcenteno — APPROVED** [review]
LGTM

**Pull request merged**